Declaration checker cleanups

slavapestov opened the pull request:
Inspired by a forum post by @davezarzycki: https://forums.swift.org/t/is-decl-semantic-checking-two-or-three-pass/11231:

- `typeCheckDecl()` no longer has `IsSecondPass`
- `validateDecl()` no longer calls `typeCheckDecl()`
- Re-declaration checking is now only done for declarations in the primary file(s)

Removing `IsFirstPass` is a bit trickier.

slavapestov committed:
Sema: Remove DeclChecker::IsSecondPass

The DeclChecker had three possible states:

- IsFirstPass true, IsSecondPass false. This is the 'first pass' for
  declarations that appear at the top-level, or are nested inside
  top-level types.

- IsFirstPass false, IsSecondPass true. This is the 'second pass' for
  declarations that appear at the top-level, or are nested inside
  top-level types.

- IsFirstPass false, IsSecondPass false. This was used for (some)
  local declarations.

This is unnecessarily confusing. We can eliminate the third state
by calling typeCheckDecl() twice in a few places. This allows
IsSecondPass to be removed entirely since it's now always equal to
!IsFirstPass.

jrose-apple commented on lib/Sema/TypeCheckStmt.cpp:
```diff
@@ -480,6 +480,7 @@ class StmtChecker : public StmtVisitor<StmtChecker, Stmt*> {
   }
     
   Stmt *visitDeferStmt(DeferStmt *DS) {
+    TC.typeCheckDecl(DS->getTempDecl(), /*isFirstPass*/true);
```
I think these are deliberately not doing two passes because they are local, as you said.

slavapestov replied:
I'm hoping to get rid of the two passes altogether, but even then if we make two passes over local declarations it shouldn't make much of a difference since it's doing the same amount of work in the end.

slavapestov committed:
ClangImporter: Don't fall back on Sema for synthesizing the memberwise init

The importer would only synthesize its own memberwise init if the
declaration had a zero initializer, which supresses Sema's automatic
synthesis. Trouble is we cannot rely on Sema synthesizing the
memberwise init in case we deserialize declarations after type
checking.

Some other changes caught this with an assert, so fix the importer
to do the right thing here.

slavapestov committed:
Sema: Skip 'non-member observing property' check for invalid properties

This avoids emitting a diagnostic twice with a subsequent change.

slavapestov committed:
Sema: Remove unnecessary enum recursiveness check

slavapestov committed:
Sema: Move most of DeclChecker::visitEnumElementDecl() to validateDecl()

slavapestov committed:
Sema: Move most of DeclChecker::visitSubscriptDecl() to validateDecl()

slavapestov committed:
Sema: Move most of DeclChecker::visitDestructorDecl() to validateDecl()

slavapestov committed:
Sema: Move most of DeclChecker::visitConstructorDecl() to validateDecl()

slavapestov committed:
Sema: Move most of DeclChecker::visitFuncDecl() to validateDecl()

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please test source compatibility

slavapestov committed:
Sema: Designated initializer overrides are fully type-checked

slavapestov committed:
Sema: Don't validate the other declaration in checkRedeclaration()

We're going to visit it later after we type check it anyway.

slavapestov commented:
@swift-ci Please smoke test

slavapestov committed:
Sema: Call setValidationStarted() on various synthesized FuncDecls

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please test source compatibility

slavapestov commented:
@swift-ci Please test source compatibility

slavapestov commented:
@swift-ci Please smoke test compiler performance

slavapestov commented:
@swift-ci Please smoke test compiler performance

aschwaighofer commented:
This causes a failure on some bots and PR testing https://ci.swift.org/job/oss-swift-pr-test/1659/

I am disabling the test until you get a chance to look at it: #15421

jrose-apple commented:
The test run you linked doesn't show that test failing. Can you find one where it does?

(The test you named has very little to do with this change.)

aschwaighofer commented:
https://ci.swift.org/job/oss-swift-incremental-RA-linux-ubuntu-14_04/3376/consoleFull#-2072434866ba62d58e-7248-467b-91e0-c7508d5cf947

aschwaighofer commented:
I will revert the disabling commit again since it only failed once on one bot: #15426

jrose-apple commented:
Yeah, that's definitely weird but it's very unlikely to be this change's fault.

slavapestov commented:
That yet has been flaky in the past. It’s not related to these changes
